### PR TITLE
BUG/DOC: Add documentation in types/common.py

### PR DESCRIPTION
**Pull request by gfyoung**
Adds documentation for all internal functions in `types/common.py`

In addition, caught a bug in which some functions calling `_get_dtype` were not catching the `TypeError`.  Documented those functions along the way too.

Partially addresses #15895.

### CODE REVIEW

File: pandas/types/common.py [hunk 1]
```diff
@@ -116,8 +144,39 @@ def is_categorical_dtype(arr_or_dtype):
 
 
 def is_string_dtype(arr_or_dtype):
-    dtype = _get_dtype(arr_or_dtype)
-    return dtype.kind in ('O', 'S', 'U') and not is_period_dtype(dtype)
+    """
+    Check whether the provided array or dtype is of the string dtype.
```
Comment (jreback) [inline pandas/types/common.py]: see comments here: https://github.com/pandas-dev/pandas/issues/15585

maybe add a reference / comments

Comment (gfyoung) [inline pandas/types/common.py]: Ah, sounds good.  Will do.

Comment (gfyoung) [inline pandas/types/common.py]: Ah, sounds good.  Done.

File: pandas/types/common.py [hunk 2]
```diff
@@ -220,10 +313,22 @@ def is_datetime_or_timedelta_dtype(arr_or_dtype):
...
+    we need to condition based on Python version.
+
+    Parameters
+    ----------
+    e : Exception or sub-class
+        The exception object to check.
+
+    Returns
+    -------
+    is_orderable_exc : Whether or not the exception raised is an
```
Comment (jreback) [inline pandas/types/common.py]: I think better to use say:

```
Returns
--------
boolean : .......
``` 

(for all of these)

File: pandas/tests/types/test_common.py [hunk 3]
```diff
@@ -80,3 +81,15 @@ def test_dtype_equal_strict():
...
         pandas_dtype('datetime64[ns, US/Eastern]'),
         pandas_dtype('datetime64[ns, CET]'))
+
+
+def test_get_dtype_error_catch():
+    # see gh-15941
+    #
+    # No exception should be raised.
+
+    assert not com.is_string_dtype(None)
```
Comment (jreback) [inline pandas/tests/types/test_common.py]: can you add *all* of these (use parametrize / construct them)

Comment (gfyoung) [inline pandas/tests/types/test_common.py]: Yep, done.

File: pandas/types/common.py (hunk 2 above)
Comment (gfyoung) [inline pandas/types/common.py]: Yep, done.

File: pandas/tests/types/test_common.py [hunk 4]
```diff
@@ -80,3 +81,22 @@ def test_dtype_equal_strict():
     assert not is_dtype_equal(
         pandas_dtype('datetime64[ns, US/Eastern]'),
         pandas_dtype('datetime64[ns, CET]'))
+
+
+get_dtype_funcs = [
+    (com.is_dtype_equal, 2),
```
Comment (jreback) [inline pandas/tests/types/test_common.py]: what I meant was to add *all* of the validation functions.

```
(In [3]: [ f for f in dir(com) if f.startswith('is_') and f.endswith('dtype')]
Out[3]: 
['is_any_int_dtype',
 'is_bool_dtype',
 'is_categorical_dtype',
 'is_complex_dtype',
 'is_datetime64_any_dtype',
 'is_datetime64_dtype',
 'is_datetime64_ns_dtype',
 'is_datetime64tz_dtype',
 'is_datetime_or_timedelta_dtype',
 'is_float_dtype',
 'is_floating_dtype',
 'is_int64_dtype',
 'is_int_or_datetime_dtype',
 'is_integer_dtype',
 'is_numeric_dtype',
 'is_object_dtype',
 'is_period_dtype',
 'is_signed_integer_dtype',
 'is_string_dtype',
 'is_string_like_dtype',
 'is_timedelta64_dtype',
 'is_timedelta64_ns_dtype',
 'is_unsigned_integer_dtype']
```

then ``is_dtype_equal`` should be a separate function to test (as its different)

Comment (jreback) [inline pandas/tests/types/test_common.py]: could make this a nice fixture :>

Comment (gfyoung) [inline pandas/tests/types/test_common.py]: Well the whole point was to check all of the functions that `_get_dtype`.  The ones in the `list` are the only ones that do.

Comment (gfyoung) [inline pandas/tests/types/test_common.py]: > could make this a nice fixture :>

Perhaps, if it's needed :smile:

Comment (jreback) [inline pandas/tests/types/test_common.py]: hmm, well we should check them all anyhow :>

Comment (gfyoung) [inline pandas/tests/types/test_common.py]: Fair enough.  Will do then.

**Comment by jreback** [comment]
thanks @gfyoung

**Pull request merged**